Property editor changes will call `OnValidate` on components

Remscar opened the pull request:
- When using the undo scope only components specifically listed in `.WithComponentChanges()` will have `Validate` called. Documentation on the undo system states that `GameObjectUndoFlags.Components` captures the _list_ of components, not changes to their contents, while `WithComponentChanges` clearly indicates it will capture changes to the components themselves.
- An edge case arises with prefab instances where `.WithComponentChanges()` is ignored when modifying a prefab instance in the scene. This is why `_capturedComponentsToValidate` is used to keep track of all components passed from the builder via `WithComponentChanges`
- Tested: Undo and Redo properly call `OnValidate` when a value changes
- In testing I discovered that pasting a component "As New" will not call `OnValidate()` so I fixed that as well (Change to `Component.DeserializeImmediately`)

There is a potentially sub-optimal behavior that needs to be fixed:

**"Revert Property Instance Change" calls `OnValidate` twice**
- Once from `Component.DeserializeImmediately` which `EditorUtility.Prefabs.RevertPropertyChange` ultimately calls
- Once from the UndoSnapshot (via the code that was introduced in this PR)

If this is a blocker then this behavior could be fixed be adding a `validate` flag to `Component.DeserializeImmediately` to conditionally call OnValidate. `Component.DeserializeImmediately` seems to only ever be called by Prefab related code.

garrynewman committed:
ControlSheetRow.Property is a [Parameter]

sboxbot committed:
Open source release

This commit imports the C# engine code and game files, excluding C++ source code.

[Source-Commit: ceb3d758046e50faa6258bc3b658a30c97743268]

handsomematt committed:
.gitignore for public repo

handsomematt committed:
Add missing shader includes

handsomematt committed:
Do sync step sooner, avoid any extra junk made during tests/upload

CarsonKompon committed:
Fix "Render Backface" toggle not working in Material made from ShaderGraph shaders (#3261)

lolleko committed:
Move InterfaceControlWidget.cs to uppercase code dir

garrynewman committed:
Add EnumControl

garrynewman committed:
Add Label.Tokenize (turn off automatic conversion of strings starting with #)

garrynewman committed:
TextEntry has tokenize turned off, strings can start with #

garrynewman committed:
Add ColorControl

CarsonKompon committed:
Fix not being able to "Paste as New Component" on prefab instances (#3432)

Resolves Facepunch/sbox-issues#9604

CarsonKompon committed:
SceneTrace uses the passed scene.PhysicsWorld instead of Game.ActiveScene.PhysicsWorld (#3433)

Resolves Facepunch/sbox-issues#9598 (Fixes issues where traces in the editor can throw NREs after exiting play mode)

solwllms committed:
Menu-in-editor fixes 2 (#3395)

* Skip menu editor code outside of the menu project
* Asset Browser: skip menu project assets in Everything and Recents

lolleko committed:
Free native engine dlls before shutdown (#3307)

handsomematt committed:
Render DebugOverlay before UI

lolleko committed:
Less aggressive shutdown during standalone test (#3438)

lolleko committed:
Fix access error when cleaning up standalone test (#3441)

Add slight delay because windows may still briefly hold a lock on some files even after the process exited

handsomematt committed:
Use a tagged commit for the public baseline, using a date jumped between different commits

aylaylay committed:
Fix tool widget when no subtools (#3435)

lolleko committed:
Refactor the public repo sync (#3444)

* Refactor sync script

Use single filter repo call
Use globs instead of python madness
Ensure shallow clone is clean before working on it

* Add dry run option to sync script
* Add public gitattributes

ryleigh committed:
Prevent Camera Preview LookAt mode from exiting immediately (#3419)

DevulTj committed:
Play mode: re-add game view resolution/ratio dropdown functionality

handsomematt committed:
Public PR triage workflow

handsomematt committed:
Public PR triage uses pull_request_target

sampavlovic committed:
Fix leaky swapchain when resizing on game mode (#3389)

* Fix leaky swapchain when resizing on game mode and dispose of handle after usage
VideoRecorder would get a strong handle copy of the swapchain every frame and retain it, causing native to fail to shutdown the copies from it
Made usage of getting native swapchain consistent on managed on other stuff, doing it like this ensures GC properly disposes of the strong handles even not disposing explicitly

Remove now unused ScreenRecorder.def and ScreenshotService.def

There are still optimizations to swapchain I'd like to send on another commit, game is allocating way more swapchains than needed even without the leak, has two completely different paths depending if you are MSAA or not, this can all be much simpler
https://files.facepunch.com/sampavlovic/1b1811b1/EjDyxbTahs.png

* Remove NativeLayerRenderTarget, was unused and fucked

* Keep it as an ITexture/HRenderTextureStrong in managed so we avoid IDisposable, ReadTextureAsync with ITexture

sampavlovic committed:
Support instanced tint on Blendable and Material::From( i ) (#3449)

* Support instanced tint on Blendable and Material::From( i )

Blendable supports vertex tint color, but was initially intended for world geometry so there was no point in working with instance color, now works fine

https://files.facepunch.com/sampavlovic/1b2611b1/sbox-dev_qkWlvBOXW8.png

* Build shaders

lolleko committed:
Public bootstrap robust download (#3450)

* More robust downloading of artifacts

Don't download to a temp file first
Try to retry download up to 3 times, if it fails
Fail Bootstrap if any download fails

* Fix contentbuilder and shadercompiler not forwarding to stdout

lolleko committed:
Move menu code from lowercase to uppercase dir to fix sync (#3451)

lolleko committed:
Fix git filtering causing inconsistent commit hashes when files are deleted

Move filter implementation to python so we have more flexibility when implementing filters.
Make sure deleted files are properly filtered by evaluating globs in the filter itself not before.
Make sure deleted LFS files are accounted for by scanning the history of the shallow clone for delete/changed lfs files.
Whitelist some additional shaders.

garrynewman committed:
Color mixer popup

Matt9440 committed:
Clicking anywhere within ContentBlockHero brings up the create game modal

sampavlovic committed:
Terrain Fixes - Seamsless & Optimizations (#3442)

* Fix terrain seams and optimize

Overlap LODs by one step to fix holes in LOD transitions
Reuse vertices that exist on same key when building diamond square

https://files.facepunch.com/sampavlovic/1b2411b1/8mb.video-eW2-tNb22a60.mp4

* Add NoTile class and make terrain use it
https://files.facepunch.com/sampavlovic/1b2411b1/sbox-dev_R1FwUmLhvu.mp4
https://files.facepunch.com/sampavlovic/1b2411b1/sbox-dev_YhKyIwvhve.mp4

* Sure why not Mr. Robot

handsomematt committed:
Revert "Render DebugOverlay before UI"

This reverts commit dd44dc6eb842cfccc430e9211f9999ff4123e104.

solwllms committed:
Restore separate GameEditorSessions (#3453)

* SceneEditorSession: make game vs editor scenes more distinct, Scene is whichever is currently active

* Route prefab update, model reload etc events to both scenes if needed

* SceneTree update checking a bit cleaner

* Bring back GameEditorSessions instead, so undo, selection etc can all be linked 1:1 with scenes again

* tweak and tidy

DevulTj committed:
Give SceneViewWidget a background so it doesn't look so weird when fullscreening at a low resolution preset (#3457)

solwllms committed:
SceneRenderingWidget: still do Scene.PreRender without a camera in the scene (#3456)

Fixes Facepunch/sbox-issues#9567

lolleko committed:
Fix native interop gen being skipped in in private build (#3462)

lolleko reviewed (changes requested):
Thanks for giving this a try we still need some more changes though.
Sorry maybe i wasn't clear enough, `OnValidate` should only be invoked by the undo system, not via widget.
It should be invoked:

1. When a component snapshot is created (this happens after a change via gizmo, inspector change, context menu change).
2. When undo is invoked and a component has it's state restored
(might not be needed since Component.Deserialize already calls on Validate)
3. When redo is invoked and a component has it's state restored
(might not be needed since Component.Deserialize already calls on Validate)

There is also the case where component changes are captured in the undo system by storing the whole GameObject `WithGameObjectChanges( go, GameObjectUndFlags.Components)` or `WithGameObjectChanges( go, GameObjectUndFlags.All)`.
I am not sure what do do in this case. Maybe we can just run `OnValidate` on all components of the Go in that case.
(might not be needed since Component.Deserialize already calls on Validate)

You should get rid of the invocation in the ComponentListWidget and only do it in the correct places in the undo system.

And finally: instead of using `EditorEvents` we usually put these helpers into `EditorUtility`.

To test this, add a log to on validate and then verify it is triggered when modifying a component via gizmos (e.g. box collider), when its changed via context menu and via inspector.

lolleko commented on engine/Sandbox.Tools/Scene/Session/SceneEditorSession.Undo.cs:
```diff
@@ -210,6 +210,15 @@ public void PostRestore( Scene scene )
+				var comp = compRef.Resolve( scene );
+				if ( comp.IsValid() )
+				{
+					comp.Validate();
```
Probably just an oversight , feel free to simplify the above loop.
I also think this could be just on loop.
Unless you want to make sure changes to all components are restored before calling OnValidate.
Honestly no strong opinion on this,

Remscar committed:
Call OnValidate on property change

Remscar commented:
Okay here is the next revision;

- When using the undo scope only components specifically listed in `.WithComponentChanges()` will have `Validate` called.
- An edge case arises with prefab instances where `.WithComponentChanges()` is ignored when modifying a prefab instance in the scene. This is why `_capturedComponentsToValidate` is used to keep track of all components passed from the builder via `WithComponentChanges`
- Tested: Undo and Redo properly call `OnValidate` when a value changes
- In testing I discovered that pasting a component "As New" will not call `OnValidate()` so I fixed that as well (Change to `Component.DeserializeImmediately`)

There is a potentially sub-optimal behavior that needs to be fixed:

**"Revert Property Instance Change" calls `OnValidate` twice**
- Once from `Component.DeserializeImmediately` which `EditorUtility.Prefabs.RevertPropertyChange` ultimately calls
- Once from the UndoSnapshot (via the code that was introduced in this PR)

If this is a blocker then this behavior could be fixed be adding a `validate` flag to `Component.DeserializeImmediately` to conditionally call OnValidate. `Component.DeserializeImmediately` seems to only ever be called by Prefab related code.

@lolleko You mention that "Component.Deserialize already calls on Validate" but I don't see that anywhere in the codebase. I only ever see OnValidate being called on Components if the game object is deserialized. Making components call Validate could be a more elegant solution than what i've implemented for the redo/undo, if you would prefer that let me know.

lolleko commented:
That's what i meant with it will be a bit more complicated. Thanks for hanging in.

I am a bit worried to touch `DeserializeImmediately` it's a public method changing it, may change behavior games rely on.
Maybe instead of adding a OnValidate call there you call it manually after "paste as new". This would also fix the issue of it being called twice in other instances.

Not a big fan of `_capturedComponentsToValidate`, but also don't see a better way right now, could you add a comment explaining why it exists?

I think the whole way how we handle component deserialization PostSerialize, OnValidate needs a make over.
But that's out of scope for this PR for now I would be happy with this for now and validate being called on editor changes again.

===

That being said the longer I think about OnValidate, I don't think it should exist.

We have `[Validate( nameof( IsValid ), "Warning Message", LogLevel.Warn )]` for properties which can handle editor validation.

I don't know how much `OnValidate` is used in game code, but it seems adding a callback called `PostSerialize` might make more sense.
